library/x509_create.c

```diff
@@ -123,6 +123,16 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
     return cur;
 }
 
+static int x509_is_char_hex(char c)
+{
+    return ('0' <= c && c <= '9') || ('a' <= c && c <= 'f') || ('A' <= c && c <= 'F');
+}
+
+static int x509_hex_to_int(char c)
+{
+    return ((c & 0x40) ? (c + 9) : c) & 0x0F;
```

Comment:
Please write a more readable version of this.

To make this readable as well as to save code size, it's probably better to combine this with the previous function (and see my earlier comments on function names)

```
static int hex_to_int(char c)
{
    return (c >= '0' && c <= '9') ? c - '0' :
           (c >= 'A' && c <= 'F') ? c - 'A' + 10 :
           (c >= 'a' && c <= 'f') ? c - 'a' + 10 : -1;
}
```

Then you can have something like the following (might be nicer to separate it out a bit, but you get the idea)

```
int n1, n2;    /* nibble values of possible hex byte */
:
if (c + 1 < end && (c1 = hex_to_int(*c)) != -1 && (c2 = hex_to_int(*(c+1))) != -1) {
    hexpair = 1;
    *(d++) = (c1 << 4) | c2;
    c++;
}
```